stdlib/public/Concurrency/Actor.cpp

```diff
@@ -2151,74 +2197,79 @@ static void swift_task_switchImpl(SWIFT_ASYNC_CONTEXT AsyncContext *resumeContex
+                       serialExecutorRef.getIdentity(),
+                       __taskExecutorRef.getIdentity());
+#endif
 
   assert(job && "no job provided");
-
   _swift_tsan_release(job);
 
-  if (executor.isGeneric()) {
-    // TODO: check the task for a flag if we need to look for task executor
+  if (serialExecutorRef.isGeneric()) {
     if (auto task = dyn_cast<AsyncTask>(job)) {
-      auto taskExecutor = task->getPreferredTaskExecutor();
-      if (taskExecutor.isDefined()) {
+      auto taskExecutorRef = task->getPreferredTaskExecutor();
+      if (taskExecutorRef.isDefined()) {
 #if SWIFT_CONCURRENCY_EMBEDDED
         swift_unreachable("task executors not supported in embedded Swift");
 #else
-        auto wtable = taskExecutor.getTaskExecutorWitnessTable();
-        auto taskExecutorObject = taskExecutor.getIdentity();
-        auto taskExecutorType = swift_getObjectType(taskExecutorObject);
-        return _swift_task_enqueueOnTaskExecutor(job, taskExecutorObject,
-                                                 taskExecutorType, wtable);
+        auto taskExecutorIdentity = taskExecutorRef.getIdentity();
+        auto taskExecutorType = swift_getObjectType(taskExecutorIdentity);
+        auto taskExecutorWtable = taskExecutorRef.getTaskExecutorWitnessTable();
+
+        return _swift_task_enqueueOnSerialAndTaskExecutor(
+            job,
+            serialExecutorRef,
+            taskExecutorIdentity, taskExecutorType, taskExecutorWtable);
 #endif // SWIFT_CONCURRENCY_EMBEDDED
       } // else, fall-through to the default global enqueue
     }
     return swift_task_enqueueGlobal(job);
   }
 
-  if (executor.isDefaultActor()) {
-    auto taskExecutor = TaskExecutorRef::undefined();
+  if (serialExecutorRef.isDefaultActor()) {
+    auto taskExecutorRef = TaskExecutorRef::undefined();
     if (auto task = dyn_cast<AsyncTask>(job)) {
-      taskExecutor = task->getPreferredTaskExecutor();
+      taskExecutorRef = task->getPreferredTaskExecutor();
     }
 
-#if SWIFT_CONCURRENCY_EMBEDDED
-    swift_unreachable("task executors not supported in embedded Swift");
-#else
-    if (taskExecutor.isDefined()) {
-      auto wtable = taskExecutor.getTaskExecutorWitnessTable();
-      auto executorObject = taskExecutor.getIdentity();
-      auto executorType = swift_getObjectType(executorObject);
-      return _swift_task_enqueueOnTaskExecutor(job, executorObject,
-                                               executorType, wtable);
-    } else {
-      return swift_defaultActor_enqueue(job, executor.getDefaultActor());
+    if (taskExecutorRef.isDefined()) {
+      SWIFT_TASK_DEBUG_LOG(
+          "enqueue job %p on default actor %p with task executor %p", job,
+          serialExecutorRef.getDefaultActor(), serialExecutorRef.getIdentity());
+      return swift_defaultActor_enqueue_withTaskExecutor(
+          job, serialExecutorRef.getDefaultActor(), taskExecutorRef);
```

Comment:
^^^ This is part of the important change -- instead of just calling `_swift_task_enqueueOnTaskExecutor` we have to do an enqueue, and respect the task executor when we actually run it -- rather than on global pool, we'll use the task executor then.